hudi-common/src/main/java/org/apache/hudi/common/model/HoodieAvroRecordMerger.java

```diff
@@ -67,12 +74,12 @@ private HoodieRecord preCombine(HoodieRecord older, HoodieRecord newer) {
   }
 
   private Option<HoodieRecord> combineAndGetUpdateValue(HoodieRecord older, HoodieRecord newer, Schema schema, Properties props) throws IOException {
-    Option<HoodieAvroIndexedRecord> previousRecordAvroPayload = older.toIndexedRecord(schema, props);
-    if (!previousRecordAvroPayload.isPresent()) {
+    Option<IndexedRecord> previousAvroData = older.toIndexedRecord(schema, props).map(HoodieAvroIndexedRecord::getData);
+    if (!previousAvroData.isPresent()) {
       return Option.empty();
     }
 
-    return ((HoodieAvroRecord) newer).getData().combineAndGetUpdateValue(previousRecordAvroPayload.get().getData(), schema, props)
+    return ((HoodieAvroRecord) newer).getData().combineAndGetUpdateValue(previousAvroData.get(), schema, props)
         .map(combinedAvroPayload -> new HoodieAvroIndexedRecord((IndexedRecord) combinedAvroPayload));
```

Comment:
@wzx140 we should actually produce `HoodieAvroRecord` here instead:

 - Merged records will be subsequently reshuffled a few times, and shuffling `IndexedRecord` will kill performance
 - Instead we should use `HoodieAvroRecord` holding some dummy `RecordPayload` just holds Avro bytes for us (and throws if anyone will try to use it for merging) 

Does it make sense?

Reply:
I don't think so. In HoodieAvroRecordMerger, we do `combineAndGetUpdateValue` before writing and shuffling is done before `combineAndGetUpdateValue`. We do `preCombine` with record which hold avro byte and then shuffle the result. So there is no performce loss. This is also the original logic.